### ISSUE TITLE
BUG FIXED! All Previous FFT Performance Are in Training Data...which is WRONG!

### ISSUE THREAD

**Issue by dichen001** [ai-se/FFT_Jack#6]
# FFT Performance on Test Data
![image](https://user-images.githubusercontent.com/15117843/33109831-ca13b890-cf11-11e7-9dba-0bb1f39480a8.png)

**Comment by timm** [comment]
ooops... self test? classic error

1. your results are consistent with mis-labelling the graphs: you sure you got accuracy and dist to heaven labelled right? not the reverse?
2. the results in the third column show many FFTs falling to the same point. this is consistent with a bug in whatever it used to assess the ranks in the third column
3. were going to need stats tests: stats.py applied to distance to heaven for the FFT results in column1,2,3,4
4. also, please offer a table showing distance to heave for NB,SMO,SL EM. this will be one number per learner per data set (where as the items in the last line will be 16*4 numbers)
5. it looks like different learners respond best to different biases . eg. infogain does well for synapse. so maybe the lesson here is use the simplest learner possible, then apply hyperparameter optimization to tune its choices
6. were going to need L0C vs Recall results .
7. still some 0,0 and 1,1 results. we'll need an explanation for that. another bug?

**Comment by dichen001** [comment]
- Reformated code and fixed the bugs. **now, consistent with the FFT paper, level 4 FFT generate 2^(4-1) = 8 trees**
1. fixed.
2. fixed.
3. not sure what tests to run, but collected all the dist2heaven for all the FFTs generated per criteria per data.
4. Check the figure below
![image](https://user-images.githubusercontent.com/15117843/33253791-50ee7cd8-d313-11e7-8610-3c273bcf0887.png)
5.  
![image](https://user-images.githubusercontent.com/15117843/33383860-73794496-d4f2-11e7-9d8c-1a4dcb0b79aa.png)



6. not exactly sure about the y-axis. now the y axis is got  by (current hits / all the bugs). So the predict may not predict all the bugs.

7. Double checked. Not sure how to explain.
![image](https://user-images.githubusercontent.com/15117843/33254141-5d613008-d315-11e7-9905-a94a2e5edc5c.png)
![image](https://user-images.githubusercontent.com/15117843/33254150-64d8c2ba-d315-11e7-9673-57473e7823f1.png)

**Comment by dichen001** [comment]
loc updated

**Comment by dichen001** [comment]
# FFT(optimized for Dist2Heaven on [Recall,FalseAlarm] ) vs FFT(optimazed for AUC of [Recall,LOC])
![image](https://user-images.githubusercontent.com/15117843/33393696-fb757bec-d50d-11e7-9c61-3d0702d5874e.png)

# Dist2Heaven Comparison on [Recall,FalseAlarm] 
![image](https://user-images.githubusercontent.com/15117843/33393815-6a0d2a8c-d50e-11e7-88de-edd724495610.png)

# P_opt(m) (The larger the better)
![image](https://user-images.githubusercontent.com/15117843/33395655-34dca7c4-d514-11e7-9d07-5f2d224f16c1.png)

![image](https://user-images.githubusercontent.com/15117843/33395631-26b6c7ba-d514-11e7-9a40-a56a198830ea.png)

**Comment by dichen001** [comment]
## Changes
- level 4 FFT generates 2^(4-1) = 8 trees, with structures as the following:
    - [[1, 1, 1], [1, 1, 0], [1, 0, 1], [1, 0, 0], [0, 1, 1], [0, 1, 0], [0, 0, 1], [0, 0, 0]]
- Best tree is selected in the **training data** instead of the **testing data**(previou version).
- Remove FFT generated by Gini and InfoGain. 
- Added Plots for LOC comparation.
- Added csv for all data. 


## Recall, False Positive
![image](https://user-images.githubusercontent.com/15117843/33641832-216b86b0-da05-11e7-9901-026e8de975cd.png)


## Bug Detection, %LOC
![image](https://user-images.githubusercontent.com/15117843/33641822-19c04072-da05-11e7-8f30-6803dc3befc3.png)


## Data all in one
Excel can be found [here](https://github.com/ai-se/FFT_Jack/blob/aafc3d5a1dca8346a52f02d3055b3daa7551fabd/data/2in1.xlsx)
![image](https://user-images.githubusercontent.com/15117843/33641490-b3e05fa4-da03-11e7-898b-d33006fba05b.png)

**Comment by dichen001** [comment]
## Changes
- FFT from 8 to 16.
- Added Frequency for Different FFT Structures.

## Recall, False Positive
![image](https://user-images.githubusercontent.com/15117843/33684802-88a65182-da9d-11e7-8965-1d9c52cdb999.png)

## Bug Detection, %LOC
![image](https://user-images.githubusercontent.com/15117843/33684795-821d0f5e-da9d-11e7-8fea-7b748510b2dd.png)

## FFT Structure Frequency
![image](https://user-images.githubusercontent.com/15117843/33684759-5a49fe6a-da9d-11e7-86ae-02703a15231a.png)

## Data all in one
New Excel can be found [here](https://github.com/ai-se/FFT_Jack/blob/master/data/2in1.xlsx)
![image](https://user-images.githubusercontent.com/15117843/33684664-043fb4e2-da9d-11e7-8fa3-d39603ee1e9c.png)

Seems like still no big differences on P_opt between FFT_LOC and FFT_ACC.
